llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -263,9 +254,70 @@ bool RepoPruning::runOnModule(Module &M) {
     }
   }
 
-  LLVM_DEBUG(dbgs() << "size of module: " << M.size() << '\n');
-  LLVM_DEBUG(dbgs() << "size of removed functions: " << NumFunctions << '\n');
-  LLVM_DEBUG(dbgs() << "size of removed variables: " << NumVariables << '\n');
+  LLVM_DEBUG(dbgs() << "Size of module: " << M.size() << '\n');
+  LLVM_DEBUG(dbgs() << "Number of removed functions: " << NumFunctions << '\n');
+  LLVM_DEBUG(dbgs() << "Number of removed variables: " << NumVariables << '\n');
 
   return Changed;
 }
+
+static bool wasPruned(const GlobalObject &GO) {
+  if (const MDNode *const MD = GO.getMetadata(LLVMContext::MD_repo_ticket)) {
+    if (const TicketNode *const TN = dyn_cast<TicketNode>(MD)) {
+      return TN->getPruned();
+    }
+  }
+  return false;
+}
+
+static bool eliminateUnreferencedAndPrunedGOs(Module &M) {
+  bool Changed = false;
+
+  // Drop the bodies of available externally functions.
+  for (Function &F : M) {
+    if (F.isDeclaration() || !F.use_empty() || !wasPruned(F))
+      continue;
+    // This will set the linkage to external
+    F.deleteBody();
+    F.removeDeadConstantUsers();
+    NumFunctions++;
+    Changed = true;
+  }
+
+  // Drop initializers of available externally global variables.
```

Comment:
This comment implies that all available-externally variable are removed, but the if condition below is more selective than that. Could it be updated to explain this loop more accurately?

Reply:
Thanks, updated.